### PR TITLE
PHOENIX-6454: Add feature to SchemaTool to get the DDL in specificati…

### PR DESCRIPTION
**Pull request by swaroopak**
…on mode (Addendum)

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   4m 23s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 15s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   9m 10s |  4.x passed  |
| +1 :green_heart: |  compile  |   1m 33s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   0m 45s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   1m  1s |  4.x passed  |
| +0 :ok: |  spotbugs  |   3m  3s |  phoenix-core in 4.x has 951 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   0m 45s |  phoenix-tools in 4.x has 5 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 12s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   5m 47s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 33s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 33s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 34s |  phoenix-core: The patch generated 16 new + 297 unchanged - 10 fixed = 313 total (was 307)  |
| -1 :x: |  checkstyle  |   0m 12s |  phoenix-tools: The patch generated 11 new + 119 unchanged - 2 fixed = 130 total (was 121)  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 2 line(s) with tabs.  |
| +1 :green_heart: |  javadoc  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   4m  9s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 137m 36s |  phoenix-core in the patch passed.  |
| -1 :x: |  unit  |   3m 26s |  phoenix-tools in the patch failed.  |
| -1 :x: |  asflicense  |   0m 19s |  The patch generated 1 ASF License warnings.  |
|  |   | 181m 54s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1233 |
| JIRA Issue | PHOENIX-6454 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux 7ffd420ce25a 4.15.0-136-generic #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / 0e2b826 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-tools.txt |
| whitespace | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/artifact/yetus-general-check/output/whitespace-tabs.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/artifact/yetus-general-check/output/patch-unit-phoenix-tools.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 5954 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-tools U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   5m 38s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +0 :ok: |  mvndep  |   5m  9s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  10m 27s |  4.x passed  |
| +1 :green_heart: |  compile  |   1m 39s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   0m 44s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   1m  2s |  4.x passed  |
| +0 :ok: |  spotbugs  |   3m 18s |  phoenix-core in 4.x has 951 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   0m 46s |  phoenix-tools in 4.x has 5 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 11s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   6m 50s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 40s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 40s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 32s |  phoenix-core: The patch generated 17 new + 294 unchanged - 13 fixed = 311 total (was 307)  |
| -1 :x: |  checkstyle  |   0m 11s |  phoenix-tools: The patch generated 17 new + 112 unchanged - 8 fixed = 129 total (was 120)  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 2 line(s) with tabs.  |
| +1 :green_heart: |  javadoc  |   1m  2s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   4m 27s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 196m 27s |  phoenix-core in the patch failed.  |
| -1 :x: |  unit  |   4m 40s |  phoenix-tools in the patch failed.  |
| -1 :x: |  asflicense  |   1m 11s |  The patch generated 1 ASF License warnings.  |
|  |   | 248m 56s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | phoenix.end2end.AuditLoggingIT |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1233 |
| JIRA Issue | PHOENIX-6454 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux cf672e60ed2f 4.15.0-142-generic #146-Ubuntu SMP Tue Apr 13 01:11:19 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / 69a9ec3 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-tools.txt |
| whitespace | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/whitespace-tabs.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/patch-unit-phoenix-core.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/patch-unit-phoenix-tools.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/testReport/ |
| asflicense | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/artifact/yetus-general-check/output/patch-asflicense-problems.txt |
| Max. process+thread count | 5163 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-tools U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/2/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-tools/src/it/java/org/apache/phoenix/schema/SchemaToolSynthesisIT.java [hunk 1]
```diff
@@ -154,18 +154,19 @@ public void testCreateIndexStatement_dropIndex() throws Exception {
     private void runAndVerify(String expected, String baseDDL) throws Exception {
         String[] arg = { "-m", "SYNTH", "-d", baseDDL };
         String result = runSchemaTool(null, arg);
+        System.out.println(result);
```
Comment (gjacoby126) [inline phoenix-tools/src/it/java/org/apache/phoenix/schema/SchemaToolSynthesisIT.java:157]: Can cut out the println.

File: phoenix-tools/src/main/java/org/apache/phoenix/schema/SchemaSynthesisProcessor.java [hunk 2]
```diff
@@ -154,7 +156,24 @@ private CreateTableStatement getCreateTableStatement(AddColumnStatement alterSta
...
-            newCreateStmt = new CreateTableStatement(createStmt, newColDef);
+            PrimaryKeyConstraint oldPKConstraint = createStmt.getPrimaryKeyConstraint();
+            List<ColumnDefInPkConstraint> pkList = new ArrayList<>();
+            for(Pair<ColumnName, SortOrder> entry : oldPKConstraint.getColumnNames()) {
+                ColumnDefInPkConstraint cd = new
+                        ColumnDefInPkConstraint(entry.getFirst(), entry.getSecond(), oldPKConstraint.isColumnRowTimestamp(entry
+                .getFirst()));
+                pkList.add(cd);
+            }
+            for(ColumnDef cd : addStmt.getColumnDefs()) {
```
Comment (gjacoby126) [inline phoenix-tools/src/main/java/org/apache/phoenix/schema/SchemaSynthesisProcessor.java:167]: What if the ALTER tries to add an existing PK column?

**Review by gjacoby126 — CHANGES_REQUESTED** [review]
Just some nits

File: phoenix-tools/src/main/java/org/apache/phoenix/schema/SchemaSynthesisProcessor.java (hunk 2 above)
Comment (swaroopak) [reply phoenix-tools/src/main/java/org/apache/phoenix/schema/SchemaSynthesisProcessor.java:167]: That's a valid point, I am thinking to create a new Jira which will validate if the statements are valid on the mini-cluster and then only proceed for synthesis. Does that sound okay?

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   4m 41s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 25s |  Maven dependency ordering for branch  |
| -1 :x: |  mvninstall  |  10m 19s |  root in 4.x failed.  |
| +1 :green_heart: |  compile  |   1m 41s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   0m 48s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   1m  7s |  4.x passed  |
| +0 :ok: |  spotbugs  |   3m 23s |  phoenix-core in 4.x has 951 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   0m 51s |  phoenix-tools in 4.x has 5 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 13s |  Maven dependency ordering for patch  |
| -1 :x: |  mvninstall  |   6m 15s |  root in the patch failed.  |
| +1 :green_heart: |  compile  |   1m 46s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 46s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 32s |  phoenix-core: The patch generated 15 new + 140 unchanged - 10 fixed = 155 total (was 150)  |
| -1 :x: |  checkstyle  |   0m 12s |  phoenix-tools: The patch generated 15 new + 119 unchanged - 2 fixed = 134 total (was 121)  |
| -1 :x: |  whitespace  |   0m  0s |  The patch 2 line(s) with tabs.  |
| +1 :green_heart: |  javadoc  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   4m 39s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 197m  3s |  phoenix-core in the patch passed.  |
| -1 :x: |  unit  |   3m 25s |  phoenix-tools in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 18s |  The patch does not generate ASF License warnings.  |
|  |   | 245m  4s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1233 |
| JIRA Issue | PHOENIX-6454 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux 81fc8b31281d 4.15.0-136-generic #140-Ubuntu SMP Thu Jan 28 05:20:47 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / eb8a91e |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/branch-mvninstall-root.txt |
| mvninstall | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/patch-mvninstall-root.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/diff-checkstyle-phoenix-tools.txt |
| whitespace | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/whitespace-tabs.txt |
| unit | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/artifact/yetus-general-check/output/patch-unit-phoenix-tools.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/testReport/ |
| Max. process+thread count | 5651 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-tools U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1233/3/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by gjacoby126 — APPROVED** [review]

**Pull request merged**